### ISSUE TITLE
Create HTTP requests IO layer

### ISSUE THREAD

**Issue by jesserobertson** [jesserobertson/cogj-spec#2]
Add reader which specialises to use get range requests to parse COGJ file given URL.

[WFS 2.0 Spec](http://docs.opengeospatial.org/is/09-025r2/09-025r2.html)

# 12/02/2019 Catch-up
Principle: The lambda should be as simple as possible. Easily documentable in a blog-post-style format with pseudocode demonstrating implementations in other languages.
- [ ] Start from a new branch using [Zappa](https://github.com/Miserlou/Zappa) + Flask
- [ ] Remove geospatial dependencies. Replace with lxml templating.
- [ ] Limit to Polygons for MVP.
- [ ] Throw out feature clipping.
- [ ] Implement WFS 2.0 feature paging to deal with the API Gateway 10mb limit issue.
- [ ] [Flask-RESTPlus](https://flask-restplus.readthedocs.io/en/stable/)

## TODO

- [ ] Convert GetCapabilities response to WFS 2.0 (_only_)
- [ ] Remove the shapely dependency
- [ ] Implement a DescribeFeatureType response
- [ ] Implement a GetFeature response (with a hard-coded feature limit to start with)
- [ ] Implement WFS 2.0 feature paging and remove hard-coded limit
- [ ] Do clients need us to implement `resultType=hits`
- [ ] Better attribute type sniffing (they're all strings at the moment)
- [ ] A better way to determine the gid/fid/et cetera
- [ ] Write tests around the GeoJSON2GML code before refactoring it
- [ ] Clean up and properly name XML templates
- [ ] Response BBOX could lazily be the dissolved bbox of the feature class bounding boxes? Probably works better for more detailed RTrees where we're reading most of the features anyway. Will this cause problems for QGIS with its complaining about layer extents not matching? Future enhancement work.

## BUGS

- [ ] Issues rendering MultiPolygons with exterior and interior rings in QGIS `fid=39714 (-76.12674,39.50170)` and `fid=97307 (-76.07833,39.55162)`
- [ ] 
- [ ] Remove `test` namespace
- [ ] Check coordinate order requirements in the GML 3.2/WFS.2 spec (and the "Layer extended reported by the server is not correct" errors that QGIS throws.)

## Stretch goals

- [ ] Add a Leaflet-based WFS playground

# TODO
- [x] Setup Flask
- [x] Setup [Geolambda](https://github.com/developmentseed/geolambda)
- [x] Make WFS 1.0.0 GetCapabilities, DescribeFeatureType, and GetFeatureInfo work (simple bounding box filters only)
- [x] Read the BBOX from the header for GetCapabilities
- [ ] Support reading our COGJ with embedded RTrees (currently it only supports the bounding box-based `collections` in the original Boundless spec)
- [ ] Send HTTP Range requests in parallel
- [ ] Support GMLifying using ogr2ogr's `-append` flag so we don't have to read the GeoJSON in and pass it around
- [ ] Work out how to deploy Geolambda + Flask
- [ ] Run OGC's [test suite](https://github.com/opengeospatial/ets-wfs10)
- [ ] [AWS API Gateway](https://docs.aws.amazon.com/apigateway/latest/developerguide/limits.html) has a response payload limit of 10MB ([hacky workaround](https://stackoverflow.com/a/51711126) - assumes the client will follow redirects)
- [ ] Make [bbox intersections faster](https://gis.stackexchange.com/questions/196074/how-to-properly-implement-a-bounding-box-for-shapely-fiona)

# Questions
- [ ] Decide how users will be able to supply service metadata (e.g. Service title, abstract)

# Enhancements
- [ ] Return exceptions as WFS 1.0.0 compliant XML errors
- [ ] We assume data is in EPSG:4326. Do we need to support reprojection?
- [ ] Use Flask's streaming capabilities to stream the response back
- [ ] [Enable HTTP2](https://github.com/developmentseed/geolambda/issues/46)
- [ ] Keep an eye on the Geolambda work to make an [Lambda Layer](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html) containing the necessary GDAL et cetera libraries. [Related PR](https://github.com/developmentseed/geolambda/pull/44) and [develop branch](https://github.com/developmentseed/geolambda/commits/develop). This is probably a pre-req for moving to [Zappa](https://github.com/Miserlou/Zappa) for deploys.
